src/System.Threading/src/System/Threading/ReaderWriterLock.cs

```diff
@@ -0,0 +1,1339 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Runtime.ConstrainedExecution;
+
+namespace System.Threading
+{
+    public sealed class ReaderWriterLock : CriticalFinalizerObject
+    {
+        private const int InvalidThreadID = -1;
+        private const ushort MaxAcquireCount = ushort.MaxValue;
+        private static readonly int DefaultSpinCount = Environment.ProcessorCount != 1 ? 500 : 0;
+
+        private static long s_mostRecentLockID;
+
+        private ManualResetEvent _readerEvent;
+        private AutoResetEvent _writerEvent;
+        private long _lockID;
+        private volatile int _state;
+        private int _writerID = InvalidThreadID;
+        private int _writerSeqNum;
+        private ushort _writerLevel;
+
+#if RWLOCK_STATISTICS
+        private int _readerEntryCount;
+        private int _readerContentionCount;
+        private int _writerEntryCount;
+        private int _writerContentionCount;
+        private int _eventsReleasedCount;
+#endif // RWLOCK_STATISTICS
+
+        public ReaderWriterLock()
+        {
+            _lockID = Interlocked.Increment(ref s_mostRecentLockID);
+            GC.SuppressFinalize(this);
+        }
+
+        ~ReaderWriterLock()
```

Comment:
> Otherwise some finializers may end up getting skipped.

Is that in the case where a derive type has a finalizer, it would end up calling to base.Finalize(), and would end up skipping this due to the compiler outputting a call to the wrong base's finalizer?

If so, is it necessary in a case like this where the type is sealed?